File: implementation/scipamato/src/test/java/ch/difty/scipamato/persistance/jooq/paper/searchorder/BooleanSearchTermEvaluatorIntegrationTest.java
package ch.difty.scipamato.persistance.jooq.paper.searchorder;

import static org.assertj.core.api.Assertions.*;

import org.jooq.Condition;
import org.junit.Test;

import ch.difty.scipamato.entity.filter.BooleanSearchTerm;
import ch.difty.scipamato.entity.filter.SearchTerm;
import ch.difty.scipamato.entity.filter.SearchTermType;
import junitparams.Parameters;

/**
 * Test class to integration test the search term and the search term evaluator.
 */

public class BooleanSearchTermEvaluatorIntegrationTest extends SearchTermEvaluatorIntegrationTest<BooleanSearchTerm> {

    @SuppressWarnings("unused")
    private Object[] booleanParameters() {
        return new Object[] {
            // @formatter:off
            new Object[] { "true", true, "fn = true" },
            new Object[] { "false", false, "fn = false" }
            // @formatter:on
        };
    }

    @Override
    protected int getSearchTermType() {
        return SearchTermType.BOOLEAN.getId();
    }

    @Override
    protected BooleanSearchTerm makeSearchTerm(String rawSearchTerm) {
        return (BooleanSearchTerm) SearchTerm.of(ID, searchTermType, SC_ID, FN, rawSearchTerm);
    }

    @Override
    protected BooleanSearchTermEvaluator getEvaluator() {
        return new BooleanSearchTermEvaluator();
    }

    @Test
    @Parameters(method = "booleanParameters")
    public void booleanTest(String rawSearchTerm, Boolean value, String condition) {
        final BooleanSearchTerm st = makeSearchTerm(rawSearchTerm);
        assertThat(st.getValue()).isEqualTo(value);

        final BooleanSearchTermEvaluator ste = getEvaluator();
        final Condition s = ste.evaluate(st);

        assertThat(s.toString()).isEqualTo(condition);
    }

}

File: core/core-persistence-jooq/src/integration-test/kotlin/ch/difty/scipamato/core/persistence/paper/searchorder/BooleanSearchTermEvaluatorIntegrationTest.kt
package ch.difty.scipamato.core.persistence.paper.searchorder

import ch.difty.scipamato.core.entity.search.BooleanSearchTerm
import ch.difty.scipamato.core.entity.search.IntegerSearchTerm
import ch.difty.scipamato.core.entity.search.SearchTerm
import ch.difty.scipamato.core.entity.search.SearchTermType
import org.amshove.kluent.shouldBeEqualTo
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import org.springframework.boot.test.autoconfigure.jooq.JooqTest
import org.testcontainers.junit.jupiter.Testcontainers

/**
 * Test class to integration test the search term and the search term evaluator.
 */
@JooqTest
@Testcontainers
open class BooleanSearchTermEvaluatorIntegrationTest : SearchTermEvaluatorIntegrationTest<BooleanSearchTerm>() {

    override val searchTermType = SearchTermType.BOOLEAN.id

    private val evaluator = BooleanSearchTermEvaluator()

    override fun makeSearchTerm(rawSearchTerm: String) =
        SearchTerm.newSearchTerm(ID, searchTermType, SC_ID, FN, rawSearchTerm) as BooleanSearchTerm

    @TestFactory
    fun booleanTests() : List<DynamicTest> = mapOf(
        "true" to BooleanExp(true, "fn = true"),
        "false" to BooleanExp(false, "fn = false"),
    ).map { (rawSearchTerm, exp) ->
        DynamicTest.dynamicTest("$rawSearchTerm -> $exp") {
            val st = makeSearchTerm(rawSearchTerm)
            st.value shouldBeEqualTo exp.value

            val ste = evaluator
            val s = ste.evaluate(st)

            s.toString() shouldBeEqualTo exp.condition
        }
    }
}

private data class BooleanExp(
    val value: Boolean,
    val condition: String,
)
